libraries/fc/src/log/zipkin_appender.cpp

```diff
@@ -0,0 +1,164 @@
+   //   std::string traceId;  // [a-f0-9]{16,32} unique id for trace, all children spans shared same id
+   //   std::string name;     // logical operation, should have low cardinality
+   //   std::string parentId; // The parent span id, or absent if root span
+   //   std::string id        // a-f0-9]{16}
+   //   int64_t     timestamp // epoch microseconds of start of span
+   //   int64_t     duration  // microseconds of span
+
+   log_context context = message.get_context();
+   fc::mutable_variant_object d = message.get_data();
+
+   fc::mutable_variant_object mvo;
+   auto i = d.find( "zipkin.traceId" );
+   if( i != d.end() && i->value().is_uint64() ) {
+      uint64_t v = i->value().as_uint64();
+      mvo( "traceId", fc::to_hex( reinterpret_cast<const char*>(&v), sizeof( v ) ) );
+      d.erase( "zipkin.traceId" );
+   }
+   i = d.find( "zipkin.id" );
+   if( i != d.end() && i->value().is_uint64() ) {
+      uint64_t v = i->value().as_uint64();
+      mvo( "id", fc::to_hex( reinterpret_cast<const char*>(&v), sizeof( v ) ) );
+      d.erase( "zipkin.id" );
+   }
+   i = d.find( "zipkin.parentId" );
+   if( i != d.end() && i->value().is_uint64() ) {
+      uint64_t v = i->value().as_uint64();
+      if( v != 0 )
+         mvo( "parentId", fc::to_hex( reinterpret_cast<const char*>(&v), sizeof( v ) ) );
+      d.erase( "zipkin.parentId" );
+   }
+   i = d.find( "zipkin.name" );
+   if( i != d.end() ) {
+      mvo( "name", i->value() );
+      d.erase( "zipkin.name" );
+   }
+   i = d.find( "zipkin.timestamp" );
+   if( i != d.end() ) {
+      mvo( "timestamp", i->value() );
+      d.erase( "zipkin.timestamp" );
+   }
+   i = d.find( "zipkin.duration" );
+   if( i != d.end() ) {
+      mvo( "duration", i->value() );
+      d.erase( "zipkin.duration" );
+   }
+
+   mvo( "localEndpoint", fc::variant_object( "serviceName", service_name ) );
+
+   mvo( "tags", d );
+
+   // /api/v2/spans takes an array of spans
+   fc::variants result;
+   result.emplace_back( mvo );
+
+   return result;
+}
+
+void zipkin_appender::log( const log_message& message ) {
+   if( !my || my->consecutive_errors > my->max_consecutive_errors )
+      return;
+
+   boost::asio::post(my->work_strand, [this, message]() {
+      my->log( message );
+   });
+}
+
+void zipkin_appender::impl::log( const log_message& message ) {
+   if( consecutive_errors > max_consecutive_errors )
+      return;
+
+   try {
+      auto deadline = fc::time_point::now() + fc::microseconds( cfg.timeout_us );
+      if( !endpoint ) {
+         endpoint = url( cfg.endpoint + cfg.path );
+         std::cout << "Connecting to zipkin: " << (std::string) *endpoint << std::endl;
+      }
+
+      http.post_sync( *endpoint, create_zipkin_variant( message, cfg.service_name ), deadline );
+
+      consecutive_errors = 0;
+   } catch( const fc::exception& e ) {
+      std::cerr << "Unable to connect to zipkin: " << cfg.endpoint + cfg.path << ", error: " << e.to_detail_string() << std::endl;
```

Comment:
can these be something like a `wlog()` instead? I know it seems weird for an appender to be... appending more logs. But it seems the zipkin_appender is intended solely for zipkin trace spans, not the default logger, if I'm understanding it right?

Reply:
Nothing would prevent someone from setting up the `default` logger from using the `zip` appender causing infinite logging.

Reply:
does that make sense though? (should it even be allowed?)

Reply:
It should not be allowed. I added logic to not allows it and used default logging in `zipkin_appender`.

Reply:
I thought it would be safe to call `wlog` and `dlog` since from a diff thread. However, the `thread.join()` causes a potential deadlock. I think the safest thing is to just go back to `std::cout` & `std::cerr`